### DISCUSSION TITLE
[Bug]: setValue drops all other properties of an default

### DISCUSSION THREAD

**Discussion by eakl** [react-hook-form/react-hook-form#6388] [Q&A] [answered]
### Version Number

7.1.0

### Codesandbox/Expo snack

https://codesandbox.io/s/setvalue-svek6?file=/src/Header.js

### Steps to reproduce

Just observe the behavior in the code

I have 3 useEffect
1. One to set the default values with `reset`
2. One to set the `value` property when the dates changes
3. One to update another field when the metric array changes using the current metric values

### Expected behaviour

`metrics.0.value` is a controlled input that changes the `value` of the `metrics.0` object.

When the `value` updates with `setValue('metrics.0.value', ...)` the other keys of `metrics.0` should stay unchanged.

Current behavior, they disappear

### What browsers are you seeing the problem on?

Chrome

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
<img width="1500" alt="Screen Shot 2021-08-30 at 6 52 00 pm" src="https://user-images.githubusercontent.com/10513364/131313359-25ebbaeb-98d7-408a-861d-7d1997fc5b96.png">

**Comment by eakl** [comment, marked as answer]
The CSB is fixed.

Can you reopen the ticket?

**Reply by bluebill1049** [reply]
> Current behavior, they disappear 

can you provide more detail as the codesandbox has quite a lot going on there?

also couple things to be covered as well: 

- did you test with the latest version?
- useController/reset/setValue all happen at the same component?

**Reply by eakl** [reply]
Locally I use v7.1.0 but the CSB is using v7.4.0 and I observe the same behavior

### Context

The main component has a **form** and a **data table**

The form has 3 controlled inputs (`useController` is used at input component level)
- Visit base value
- Visit growth rate
- Click rate

The data table looks like this
```
| Metric | Jan     | Feb     | Mar    | Apr     | Mai     | Jun     | Jul     | Aug     | Sep     | Oct     | Nov     | Dec     |
+--------+---------+---------+--------+---------+---------+---------+---------+---------+---------+---------+---------+---------+
| Visits | 1000.00 | 1050.00 | 1102.5 | 1157.63 | 1215.52 | 1276.30 | 1340.12 | 1407.13 | 1477.49 | 1551.36 | 1628.93 | 1710.38 |
| Click  | 0.1     | 0.1     | 0.1    | 0.1     | 0.1     | 0.1     | 0.1     | 0.1     | 0.1     | 0.1     | 0.1     | 0.1     |
| Stream | 100.00  | 105.00  | 110.25 | 115.76  | 121.55  | 127.63  | 134.01  | 140.71  | 147.75  | 155.14  | 162.90  | 171.04  |
```

The form is responsible for
1. Setting the defaults values on mount (with `reset`)
```javascript
const defaults = {
  startDate: '2021-01-01',
  endDate: '2021-12-01',
  metrics: [
    {
      type: 'visits',
      baseValue: 1000,
      growthRate: 0.05,
      data: [0,0,0,0,0,0,0,0,0,0]
    },
    {
      type: 'clicks',
      baseValue: 0.1,
      growthRate: 0,
      data: [0,0,0,0,0,0,0,0,0,0]
    }
  ],
  final: [0,0,0,0,0,0,0,0,0,0]
}

useEffect(() => {
  reset(defaults)
}, [reset])
```

2. Watching inputs
```javascript
const startDate = useWatch({ name: 'startDate', control }) // To resize the array when this changes
const endDate = useWatch({ name: 'endDate', control }) // To resize the array when this changes
const metrics = useWatch({ name: 'metrics', control }) // To update the final array when metrics changes
```

3.  Resizing and recalculating the `data` arrays on `startDate`/ `endDate` change
```javascript
useEffect(() => {
  if (startDate && endDate) {
    for (const [i, metric] of metrics.entries()) {
      const newData = calculateNewDataArray(...)
      setValue(`metrics.${i}.data`, newData)
    }
  }
}, [startDate, endDate])
```
4. Recalculating the `final` array when one of the metric has changed
```javascript
useEffect(() => {
  if (metrics) {
    const newStream = calculateNewFinalArray(...)
    setValue(`final`, newStream)
  }
}, [metrics])
```

At `3.`, when `setValue(‘metrics.${i}.data’, newData)` happens, other keys in `metrics.${i}` disappears
I believe this is because at the input level, useController register only the input value `metrics.0.value`. setValue replaces the object instead of merging with defaults.

**Reply by bluebill1049** [reply]
The latest version is 7.14.0.

**Reply by bluebill1049** [reply]
> 3., when setValue(‘metrics.${i}.data’, newData) happens, other keys in metrics.${i} disappears
I believe this is because at the input level, useController register only the input value metrics.0.value. setValue replaces the object instead of merging with defaults.

if your input is object, we don't merge values, you need to merged and set input value with the expected result.
